src/backend/gporca/libgpos/include/gpos/memory/CMemoryPoolManager.h

```diff
@@ -117,7 +119,9 @@ namespace gpos
 				// instantiate manager
 				GPOS_TRY
 				{
-					m_memory_pool_mgr = GPOS_NEW(internal) ManagerType(internal, EMemoryPoolTracker);
+					// would've been nice if we could std::make_unique
+					auto manager = ::new ManagerType(internal, EMemoryPoolTracker);
```

Comment:
```suggestion
					ManagerType *manager = ::new ManagerType(internal, EMemoryPoolTracker);
```

> As far as I understand, auto makes the code writer's life easier at the expense of any reader, who will have to figure out the type some other way

I am open to changing _this instance_ to an explicit type. That said, I think there are more reasons than "we can be lazy" to consider using `auto`. For example:

* generic / templated code where the type is nested and even hard to read when explicitly spelled out.
* better performance and less error-prone. Handwritten-typed initialization might trip and incur a conversion:

  ```C++
  struct R {};
  struct S { S(const R&) { /* ... */} };

  // this once returned an S
  R g(int) { /* ... */ }
  void f() {
    S s = g(42);
    // auto s = g(42);
  ```

  * For more references for the considerations on `auto` and smart pointers, a good starter is Sutter's essay https://herbsutter.com/elements-of-modern-c-style/

Reply:
For the record, I think wholesale banning `auto` would be irrational, just as wholesale promoting `auto` everywhere. Informing the team on the pros and cons and reaching an informed consensus on when the pros outweigh the cons is more appropriate going forward

Reply:
While I can see the benefits of STL-type pointers and collections, I really see mostly disadvantages for `auto` in an environment where many people collaborate and the code is long-lived. If I write a program just for myself, `auto` is great. Can you tell me why that is irrational?

Reply:
Maybe we can look at it from the other side: What are the advantages of `auto`. The main one I know of is that the author doesn't need to know that the type of the variable is. Are there others? I assume it can potentially make refactoring easier, but that also potentially increases bugs when doing the refactoring. I really hope that others can chime in on this. If enough people would like to use `auto` then I'll be fine with accepting it in PRs.

Reply:
Sorry, didn't see your earlier comment above about the benefits of `auto`. To be honest, I'm just not convinced. In your example, it usually makes a difference whether s should be an R or an S, so I really don't understand how `auto` helps this particular example - quite the contrary. Generally, I would hope that we avoid ambiguous and complex C++ features. I would like our code to be as simple as possible. Postgres shows this principle very well, as far as I have seen. But, again, this is not a discussion two people should have, it's a question of how we see this as a community.

Reply:
Agreed that this should be a broader discussion. I also updated the code snippet with more comments. Regardless of the outcome of this discussion, I'd wholeheartedly recommend Sutter's short essay I linked above.

Reply:
I stumbled upon this thread.  I think there's another use case for `auto` that Jesse alluded to "universal reference" (or "forward reference") [1].  I think move semantics can be kind of tricky and `auto` helps manage that via perfect forwarding.

[1] https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers